File: 988/988.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/988.py

fun frogSum(aIn: Long, bIn: Long): Long {
    if (aIn <= 0L || bIn <= 0L) {
        throw IllegalArgumentException("a and b must be positive")
    }
    var a = aIn
    var b = bIn
    if (a > b) {
        val tmp = a; a = b; b = tmp
    }
    if (a == 1L) {
        return 0L
    }
    val width = b - 1L
    val h = LongArray((width + 1).toInt())
    for (i in 1..width.toInt()) {
        h[i] = (a * b - a * i - 1L) / b
    }
    var dp: MutableMap<Long, Pair<Long, Long>> = mutableMapOf()
    for (t in 0L..h[1]) {
        dp[t] = Pair(1L, 0L)
    }
    for (i in 2..width.toInt()) {
        val nextDp: MutableMap<Long, Pair<Long, Long>> = mutableMapOf()
        for ((prevHeight, value) in dp) {
            val (count, total) = value
            val limit = minOf(prevHeight, h[i])
            for (curHeight in 0L..limit) {
                var add = 0L
                if (prevHeight > curHeight && prevHeight > 0L) {
                    add = a * b - a * (i.toLong() - 1L) - b * prevHeight
                }
                val (oldCount, oldTotal) = nextDp.getOrDefault(curHeight, Pair(0L, 0L))
                nextDp[curHeight] = Pair(
                    oldCount + count,
                    oldTotal + total + count * add
                )
            }
        }
        dp = nextDp
    }
    var answer = 0L
    val lastColumn = width
    for ((prevHeight, value) in dp) {
        val (count, total) = value
        var add = 0L
        if (prevHeight > 0L) {
            add = a * b - a * lastColumn - b * prevHeight
        }
        answer += total + count * add
    }
    return answer
}

fun solve() {
    check(frogSum(3L, 5L) == 23L) { "frogSum(3,5) should be 23" }
    check(frogSum(5L, 13L) == 16336L) { "frogSum(5,13) should be 16336" }
    println(frogSum(19L, 53L))
}

fun main() {
    solve()
}

File: 988/988.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/988.py

use std::collections::HashMap;

fn frog_sum(mut a: i64, mut b: i64) -> i64 {
    if a <= 0 || b <= 0 {
        panic!("a and b must be positive");
    }
    if a > b {
        std::mem::swap(&mut a, &mut b);
    }
    if a == 1 {
        return 0;
    }
    let width = b - 1;
    let mut h = vec![0i64; (width + 1) as usize];
    for i in 1..=(width as usize) {
        h[i] = (a * b - a * i as i64 - 1) / b;
    }
    let mut dp: HashMap<i64, (i64, i64)> = HashMap::new();
    for t in 0..=(h[1]) {
        dp.insert(t, (1, 0));
    }
    for i in 2..=(width as usize) {
        let mut next_dp: HashMap<i64, (i64, i64)> = HashMap::new();
        for (&prev_height, &(count, total)) in &dp {
            let limit = prev_height.min(h[i]);
            for cur_height in 0..=(limit) {
                let mut add = 0i64;
                if prev_height > cur_height && prev_height > 0 {
                    add = a * b - a * (i as i64 - 1) - b * prev_height;
                }
                let (old_count, old_total) = next_dp.get(&cur_height).copied().unwrap_or((0, 0));
                next_dp.insert(
                    cur_height,
                    (old_count + count, old_total + total + count * add),
                );
            }
        }
        dp = next_dp;
    }
    let mut answer = 0i64;
    let last_column = width;
    for (&prev_height, &(count, total)) in &dp {
        let mut add = 0i64;
        if prev_height > 0 {
            add = a * b - a * last_column - b * prev_height;
        }
        answer += total + count * add;
    }
    answer
}

fn solve() {
    assert_eq!(frog_sum(3, 5), 23);
    assert_eq!(frog_sum(5, 13), 16336);
    println!("{}", frog_sum(19, 53));
}

fn main() {
    solve();
}
